executor/executor_test.go

```diff
@@ -1022,6 +1022,92 @@ func (s *testSuite) TestJSON(c *C) {
 	result.Check(testkit.Rows(`3 {} <nil>`))
 }
 
+func (s *testSuite) TestGeneratedColumnDDL(c *C) {
+	defer func() {
+		s.cleanEnv(c)
+		testleak.AfterTest(c)()
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	// check create table with generated column.
+	tk.MustExec(`CREATE TABLE test_json_ddl(a int, b int as (a+8) virtual)`)
+
+	// check desc table with generated column.
+	result := tk.MustQuery(`DESC test_json_ddl`)
+	result.Check(testkit.Rows(`a int(11) YES  <nil> `, `b int(11) YES  <nil> VIRTUAL GENERATED`))
+
+	// check show create table with generated column.
+	result = tk.MustQuery(`show create table test_json_ddl`)
+	result.Check(testkit.Rows(
+		"test_json_ddl CREATE TABLE `test_json_ddl` (\n  `a` int(11) DEFAULT NULL,\n  `b` int(11) GENERATED ALWAYS AS (a+8) VIRTUAL DEFAULT NULL\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin",
+	))
+
+	// check alter table add a generated column.
+	tk.MustExec(`alter table test_json_ddl add column c int as (b+2) stored`)
+	result = tk.MustQuery(`DESC test_json_ddl`)
+	result.Check(testkit.Rows(`a int(11) YES  <nil> `, `b int(11) YES  <nil> VIRTUAL GENERATED`, `c int(11) YES  <nil> STORED GENERATED`))
+
+	// check drop columns dependent by other column.
+	_, err := tk.Exec(`alter table test_json_ddl drop column a`)
+	c.Assert(err, NotNil)
+	terr := errors.Trace(err).(*errors.Err).Cause().(*terror.Error)
```

Comment:
I prefer expose those error in `ddl`, and use `terror.ErrorEqual(err1, err2 error) bool` for compare.
This code to get error code is too awkward here.